v0.14.3: panic: value is marked, so must be unmarked first

### Terraform Version

```
0.14.3
```

### Terraform Configuration Files
Unfortunately, it's unclear which part of the configuration is causing this. However, it's apparent that this is a pervasive issue with the latest version of Terraform, so I'm adding this to give you more info.

### Debug Output
Full trace is here: https://gist.github.com/dancrumb/26053878427f2f626d2ca4f89d951d06

Here's the Panic output:
```
Terraform v0.14.3
Configuring remote state backend...
Initializing Terraform configuration...


!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
...
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
panic: value is marked, so must be unmarked first

goroutine 614 [running]:
github.com/zclconf/go-cty/cty.Value.assertUnmarked(...)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/marks.go:123
github.com/zclconf/go-cty/cty.Value.AsBigFloat(0x2c5c280, 0xc00005442f, 0x2551bc0, 0xc000cf8400, 0x2c5c201)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/value_ops.go:1269 +0x4f
github.com/zclconf/go-cty/cty/gocty.fromCtyNumber(0x2c5c280, 0xc00005442f, 0x2551bc0, 0xc000cf8400, 0x237acc0, 0xc001bc5058, 0x182, 0xc001fbb470, 0x0, 0x0, ...)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/gocty/out.go:124 +0x4d
github.com/zclconf/go-cty/cty/gocty.fromCtyValue(0x2c5c280, 0xc00005442f, 0x2551bc0, 0xc000cf8400, 0x22f8120, 0xc001bc5058, 0x16, 0xc001fbb470, 0x0, 0x0, ...)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/gocty/out.go:85 +0x965
github.com/zclconf/go-cty/cty/gocty.FromCtyValue(0x2c5c280, 0xc00005442f, 0x2551bc0, 0xc000cf8400, 0x22f8120, 0xc001bc5058, 0x2c5c280, 0xc00005442f)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/gocty/out.go:43 +0x114
github.com/hashicorp/terraform/terraform.(*EvalValidateResource).validateCount(0xc002385bc0, 0x2c97980, 0xc000cddba0, 0x2c5ae00, 0xc0002ba780, 0xc0009bed80, 0x7f5b407707a8, 0x8)
	/home/circleci/project/project/terraform/eval_validate.go:523 +0x74c
github.com/hashicorp/terraform/terraform.(*EvalValidateResource).Validate(0xc001fbbbc0, 0x2c97980, 0xc000cddba0, 0x0, 0x0)
	/home/circleci/project/project/terraform/eval_validate.go:339 +0x16cc
github.com/hashicorp/terraform/terraform.(*NodeValidatableResource).Execute(0xc0003d66f0, 0x2c97980, 0xc000cddba0, 0xc00003a004, 0x24aa880, 0x275b780)
	/home/circleci/project/project/terraform/node_resource_validate.go:55 +0x2f5
github.com/hashicorp/terraform/terraform.(*ContextGraphWalker).Execute(0xc00149ad00, 0x2c97980, 0xc000cddba0, 0x7f5b4060bd90, 0xc0003d66f0, 0x0, 0x0, 0x0)
	/home/circleci/project/project/terraform/graph_walk_context.go:127 +0xbc
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x275b780, 0xc0003d66f0, 0x0, 0x0, 0x0)
	/home/circleci/project/project/terraform/graph.go:59 +0x962
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc0005e51a0, 0x275b780, 0xc0003d66f0, 0xc0009dbb80)
	/home/circleci/project/project/dag/walk.go:387 +0x375
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/home/circleci/project/project/dag/walk.go:309 +0x1246
```


### Crash Output
This is running on Terraform cloud, so the crash.log file is unavailable

### Expected Behavior
No crash

### Actual Behavior
Crash

### Steps to Reproduce
`terraform plan`

### Additional Context
I'm opening this as I have yet to see an issue where `AsBigFloat` is the function causing the panic.

### References

- #27107
- #27220 
- #27329

--- a/terraform/node_resource_validate.go
+++ b/terraform/node_resource_validate.go
@@ -11,8 +11,6 @@ import (
 	"github.com/hashicorp/terraform/provisioners"
 	"github.com/hashicorp/terraform/tfdiags"
 	"github.com/zclconf/go-cty/cty"
-	"github.com/zclconf/go-cty/cty/convert"
-	"github.com/zclconf/go-cty/cty/gocty"
 )
 
 // NodeValidatableResource represents a resource that is used for validation
@@ -410,69 +408,16 @@ func (n *NodeValidatableResource) validateResource(ctx EvalContext) tfdiags.Diag
 	return diags
 }
 
-func validateCount(ctx EvalContext, expr hcl.Expression) tfdiags.Diagnostics {
-	if expr == nil {
-		return nil
-	}
-
-	var diags tfdiags.Diagnostics
-
-	countVal, countDiags := ctx.EvaluateExpr(expr, cty.Number, nil)
-	diags = diags.Append(countDiags)
-	if diags.HasErrors() {
-		return diags
-	}
-
-	if countVal.IsNull() {
-		diags = diags.Append(&hcl.Diagnostic{
-			Severity: hcl.DiagError,
-			Summary:  "Invalid count argument",
-			Detail:   `The given "count" argument value is null. An integer is required.`,
-			Subject:  expr.Range().Ptr(),
-		})
-		return diags
-	}
-
-	var err error
-	countVal, err = convert.Convert(countVal, cty.Number)
-	if err != nil {
-		diags = diags.Append(&hcl.Diagnostic{
-			Severity: hcl.DiagError,
-			Summary:  "Invalid count argument",
-			Detail:   fmt.Sprintf(`The given "count" argument value is unsuitable: %s.`, err),
-			Subject:  expr.Range().Ptr(),
-		})
-		return diags
-	}
-
+func validateCount(ctx EvalContext, expr hcl.Expression) (diags tfdiags.Diagnostics) {
+	val, countDiags := evaluateCountExpressionValue(expr, ctx)
 	// If the value isn't known then that's the best we can do for now, but
-	// we'll check more thoroughly during the plan walk.
-	if !countVal.IsKnown() {
+	// we'll check more thoroughly during the plan walk
+	if !val.IsKnown() {
 		return diags
 	}
 
-	// If we _do_ know the value, then we can do a few more checks here.
-	var count int
-	err = gocty.FromCtyValue(countVal, &count)
-	if err != nil {
-		// Isn't a whole number, etc.
-		diags = diags.Append(&hcl.Diagnostic{
-			Severity: hcl.DiagError,
-			Summary:  "Invalid count argument",
-			Detail:   fmt.Sprintf(`The given "count" argument value is unsuitable: %s.`, err),
-			Subject:  expr.Range().Ptr(),
-		})
-		return diags
-	}
-
-	if count < 0 {
-		diags = diags.Append(&hcl.Diagnostic{
-			Severity: hcl.DiagError,
-			Summary:  "Invalid count argument",
-			Detail:   `The given "count" argument value is unsuitable: count cannot be negative.`,
-			Subject:  expr.Range().Ptr(),
-		})
-		return diags
+	if countDiags.HasErrors() {
+		diags = diags.Append(countDiags)
 	}
 
 	return diags

--- a/terraform/node_resource_validate_test.go
+++ b/terraform/node_resource_validate_test.go
@@ -202,6 +202,7 @@ func TestNodeValidatableResource_ValidateResource_managedResource(t *testing.T) 
 }
 
 func TestNodeValidatableResource_ValidateResource_managedResourceCount(t *testing.T) {
+	// Setup
 	mp := simpleMockProvider()
 	mp.ValidateResourceTypeConfigFn = func(req providers.ValidateResourceTypeConfigRequest) providers.ValidateResourceTypeConfigResponse {
 		if got, want := req.TypeName, "test_object"; got != want {
@@ -214,35 +215,54 @@ func TestNodeValidatableResource_ValidateResource_managedResourceCount(t *testin
 	}
 
 	p := providers.Interface(mp)
-	rc := &configs.Resource{
-		Mode:  addrs.ManagedResourceMode,
-		Type:  "test_object",
-		Name:  "foo",
-		Count: hcltest.MockExprLiteral(cty.NumberIntVal(2)),
-		Config: configs.SynthBody("", map[string]cty.Value{
-			"test_string": cty.StringVal("bar"),
-		}),
-	}
-	node := NodeValidatableResource{
-		NodeAbstractResource: &NodeAbstractResource{
-			Addr:             mustConfigResourceAddr("test_foo.bar"),
-			Config:           rc,
-			ResolvedProvider: mustProviderConfig(`provider["registry.terraform.io/hashicorp/aws"]`),
-		},
-	}
 
 	ctx := &MockEvalContext{}
 	ctx.installSimpleEval()
 	ctx.ProviderSchemaSchema = mp.GetSchemaReturn
 	ctx.ProviderProvider = p
 
-	diags := node.validateResource(ctx)
-	if diags.HasErrors() {
-		t.Fatalf("err: %s", diags.Err())
+	tests := []struct {
+		name  string
+		count hcl.Expression
+	}{
+		{
+			"simple count",
+			hcltest.MockExprLiteral(cty.NumberIntVal(2)),
+		},
+		{
+			"marked count value",
+			hcltest.MockExprLiteral(cty.NumberIntVal(3).Mark("marked")),
+		},
 	}
 
-	if !mp.ValidateResourceTypeConfigCalled {
-		t.Fatal("Expected ValidateResourceTypeConfig to be called, but it was not!")
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			rc := &configs.Resource{
+				Mode:  addrs.ManagedResourceMode,
+				Type:  "test_object",
+				Name:  "foo",
+				Count: test.count,
+				Config: configs.SynthBody("", map[string]cty.Value{
+					"test_string": cty.StringVal("bar"),
+				}),
+			}
+			node := NodeValidatableResource{
+				NodeAbstractResource: &NodeAbstractResource{
+					Addr:             mustConfigResourceAddr("test_foo.bar"),
+					Config:           rc,
+					ResolvedProvider: mustProviderConfig(`provider["registry.terraform.io/hashicorp/aws"]`),
+				},
+			}
+
+			diags := node.validateResource(ctx)
+			if diags.HasErrors() {
+				t.Fatalf("err: %s", diags.Err())
+			}
+
+			if !mp.ValidateResourceTypeConfigCalled {
+				t.Fatal("Expected ValidateResourceTypeConfig to be called, but it was not!")
+			}
+		})
 	}
 }
 

